src/main/java/com/fasterxml/jackson/core/JsonGenerator.java

```diff
@@ -2491,13 +2491,17 @@ public void copyCurrentEvent(JsonParser p) throws IOException
         }
         case ID_NUMBER_FLOAT:
         {
-            NumberType n = p.getNumberType();
-            if (n == NumberType.BIG_DECIMAL) {
-                writeNumber(p.getDecimalValue());
-            } else if (n == NumberType.FLOAT) {
-                writeNumber(p.getFloatValue());
+            Number n = p.getNumberValueExact();
+            if (n instanceof Float) {
+                writeNumber((Float) n);
+            } else if (n instanceof BigInteger) {
+                writeNumber((BigInteger) n);
+            } else if (n instanceof BigDecimal) {
+                final BigDecimal bd = (BigDecimal) n;
+                p.streamReadConstraints().validateBigIntegerScale(bd.scale());
```

Comment:
.... huh?

Reply:
in the 'failing' test case, we end up with getNumberExact returning this exact case (a BigDecimal)

Reply:
`BigDecimal` makes sense, that validation is what I don't understand (as we are not converting).

Reply:
ok, I've removed the validation